javascript/ql/src/semmle/javascript/Regexp.qll

```diff
@@ -484,3 +485,27 @@ class RegExpParseError extends Error, @regexp_parse_error {
     result = getMessage()
   }
 }
+
+/**
+ * Holds if `source` may be interpreted as a regular expression.
+ */
+predicate isInterpretedAsRegExp(DataFlow::Node source) {
+  // The first argument to an invocation of `RegExp` (with or without `new`).
+  source = DataFlow::globalVarRef("RegExp").getAnInvocation().getArgument(0)
+  or
+  // The argument of a call that coerces the argument to a regular expression.
+  exists(MethodCallExpr mce, string methodName |
+    mce.getReceiver().analyze().getAType() = TTString() and
+    mce.getMethodName() = methodName
+    |
+    (methodName = "match" and source.asExpr() = mce.getArgument(0) and mce.getNumArgument() = 1)
+    or
+    (
+      methodName = "search" and
+      source.asExpr() = mce.getArgument(0) and
+      mce.getNumArgument() = 1 and
+      // `String.prototype.search` returns a number, so exclude chained accesses
```

Comment:
This comment is a bit cryptic as it stands. You may want to explain that `search` is a common method name, and so we exclude chained accesses, which are very unlikely to be calls to `String.prototype.search`.